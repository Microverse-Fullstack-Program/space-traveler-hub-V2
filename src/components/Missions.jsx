import { useSelector, useDispatch } from 'react-redux';
import { useEffect } from 'react';
import Table from 'react-bootstrap/Table';
import { fetchMissions } from '../redux/missions/missionsSlice';
import MissionsPage from '../pages/MissionsPage';

const Missions = () => {
  const { missions } = useSelector((state) => state.missions);
  const dispatch = useDispatch();

  useEffect(() => {
    if (missions.length === 0) dispatch(fetchMissions());
  }, [dispatch, missions.length]);

  return (
    <div className="table-wrapper">
      <Table striped bordered>
        <thead>
          <tr>
            <th>#</th>
            <th>Mission</th>
            <th>Description</th>
            <th>Status</th>
            <th> </th>
          </tr>
        </thead>
        <tbody>
          {missions && missions.map((mission, index) => (
            <MissionsPage
              key={mission.id}
              mission={mission}
              index={index}
            />
          ))}
        </tbody>
      </Table>
    </div>
  );
};

export default Missions;
