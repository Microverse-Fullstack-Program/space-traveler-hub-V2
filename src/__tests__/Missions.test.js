import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import renderWithProviders from '../utils/test-utils';
import Missions from '../components/Missions';
import setupStore from '../redux/store';
import server from '../mocks/server';

describe('Missions snapshot', () => {
  test('should render missions data correctly', () => {
    renderWithProviders(
      <Provider store={setupStore({})}>
        <Missions />
      </Provider>,
    );
    const missions = document.querySelector('.table-wrapper');
    expect(missions).toMatchSnapshot();
  });
});

beforeAll(() => server.listen({ onUnhandledRequest: 'bypass' }));
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe('Test missions before fetching API', () => {
  test('Test for Loading... state', () => {
    renderWithProviders(<Missions />);
    expect(screen.getByText(/Loading.../i)).toBeInTheDocument();
    expect(screen.queryByText(/Join Mission/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/Thaicom/i)).not.toBeInTheDocument();
  });
});

describe('Test rockets after fetching API', () => {
  test('Test if rockets are rendered after fetching', async () => {
    renderWithProviders(<Missions />);
    expect(await screen.findAllByText(/Join Mission/i)).toHaveLength(10);
    expect(await screen.findAllByText(/Not A Member/i)).toHaveLength(10);
    expect(await screen.findAllByText(/Thaicom/i)).toHaveLength(2);
    expect(screen.queryByText(/Loading.../i)).not.toBeInTheDocument();
  });
});
