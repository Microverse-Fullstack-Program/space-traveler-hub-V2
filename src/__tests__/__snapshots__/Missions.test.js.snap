// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions snapshot should render missions data correctly 1`] = `
<div
  class="table-wrapper"
>
  <table
    class="table table-striped table-bordered"
  >
    <thead>
      <tr>
        <th>
          #
        </th>
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
        <th>
           
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
  <p>
    Loading...
  </p>
</div>
`;
